backends/backend_openssl3.c

```diff
@@ -2472,3 +2472,627 @@ static void openssl_ecdsa_backend(void)
+	}
+	if (dmq1 != NULL) {
+		dmq1_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(dmq1, dmq1_len, dmq1_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_EXPONENT2, dmq1_bn);
+	}
+	if (iqmp != NULL) {
+		iqmp_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(iqmp, iqmp_len, iqmp_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_COEFFICIENT1, iqmp_bn);
+	}
+	if(p!=NULL && q!=NULL){
+		BIGNUM *d_bn = BN_dup(n_bn);
+		BN_sub(d_bn, d_bn, p_bn);
+		BN_sub(d_bn, d_bn, q_bn);
+		BN_add_word(d_bn, 1);
+		BN_mod_inverse(d_bn, e_bn, d_bn, bn_ctx);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_D, d_bn);
+	}
+
+	params = OSSL_PARAM_BLD_to_param(bld);
+	ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", "fips=yes");
+	EVP_PKEY_fromdata_init(ctx);
+	ret = EVP_PKEY_fromdata(ctx, pkey, EVP_PKEY_KEYPAIR, params);
+	if(ret == 0)
+		logger(LOGGER_DEBUG, "Keygen failed\n");
+
+	return ret;
+}
+
+/************************************************
+ * CMAC/HMAC cipher interface functions
+ ************************************************/
+@@ -2569,3 +2692,498 @@ static void openssl_ecdsa_backend(void)
+{
+	register_ecdsa_impl(&openssl_ecdsa);
+}
+
+
+/************************************************
+ * RSA interface functions
+ ************************************************/
+
+#define RSA 		EVP_PKEY
+#define RSA_free(a)		EVP_PKEY_free(a)
+#define EVP_DigestSignInit(a,b,c,d,e)		EVP_DigestSignInit_ex(a,b,EVP_MD_name(c),NULL,NULL,e,NULL)
+#define EVP_DigestVerifyInit(a,b,c,d,e)		EVP_DigestVerifyInit_ex(a,b,EVP_MD_name(c),NULL,NULL,e,NULL)
+
+static int openssl_rsa_keygen_prime(struct rsa_keygen_prime_data *data,
+					flags_t parsed_flags)
+{
+	BIGNUM *e = NULL, *p = NULL, *q = NULL;
+	EVP_PKEY_CTX *ctx = NULL;
+	EVP_PKEY *pkey = NULL;
+	OSSL_PARAM_BLD *bld = NULL;
+	OSSL_PARAM *params = NULL;
+	int ret = 0;
+
+	(void)parsed_flags;
+
+	if (!data->e.len) {
+		logger(LOGGER_WARN, "RSA E missing\n");
+		return -EINVAL;
+	}
+
+	bld = OSSL_PARAM_BLD_new();
+	logger_binary(LOGGER_DEBUG, data->e.buf, data->e.len, "e");
+	logger_binary(LOGGER_DEBUG, data->p.buf, data->p.len, "p");
+	logger_binary(LOGGER_DEBUG, data->q.buf, data->q.len, "q");
+
+	e = BN_bin2bn((const unsigned char *) data->e.buf, (int)data->e.len, e);
+	CKNULL(e, -ENOMEM);
+
+	p = BN_bin2bn((const unsigned char *) data->p.buf, (int)data->p.len, p);
+	CKNULL(p, -ENOMEM);
+	if (BN_is_zero(p))
+		BN_one(p);
+	q = BN_bin2bn((const unsigned char *)data->q.buf, (int)data->q.len, q);
+	CKNULL(q, -ENOMEM);
+	OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, e);
+	if (p != NULL) {
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR1, p);
+	}
+	if (q != NULL) {
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR2, q);
+	}
+	params = OSSL_PARAM_BLD_to_param(bld);
+	ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
+	EVP_PKEY_keygen_init(ctx);
+	ret = EVP_PKEY_CTX_set_params(ctx, params);
+	data->keygen_success = 1;
+
+	if(EVP_PKEY_generate(ctx, &pkey) <= 0)
+		data->keygen_success = 0;
```

Comment:
Can we make all these conditional statements under one if with || instead of repeating assignments to "data->keygen_success"?